pybamm/solvers/base_solver.py

```diff
@@ -1162,28 +1166,72 @@ def step(
                 raise pybamm.ModelError(
                     "Cannot step empty model, use `pybamm.DummySolver` instead"
                 )
+        if non_linear_time is not None:
+            # Make sure dt is greater than the offset
+            step_start_offset = pybamm.settings.step_start_offset
+            t_start = old_solution.t[-1]
+            t_end = t_start
+            t_eval = np.array([t_start])
+            if t_start == 0:
+                t_start_shifted = t_start
+            else:
+                # offset t_start by t_start_offset (default 1 ns)
+                # to avoid repeated times in the solution
+                # from having the same time at the end of the previous step and
+                # the start of the next step
+                t_start_shifted = t_start + step_start_offset
+                t_eval[0] = t_start_shifted
+            for i in range(len(non_linear_time) - 1):
+                dt = non_linear_time[i + 1] - non_linear_time[i]
+                if dt <= step_start_offset:
+                    raise pybamm.SolverError(
+                        f"Step time must be at least {pybamm.TimerTime(step_start_offset)}"
+                    )
 
-        # Make sure dt is greater than the offset
-        step_start_offset = pybamm.settings.step_start_offset
-        if dt <= step_start_offset:
-            raise pybamm.SolverError(
-                f"Step time must be at least {pybamm.TimerTime(step_start_offset)}"
-            )
+                t_end = t_end + dt
+                # Append to t_eval
+                t_eval = np.append(t_eval, t_end)
+        elif dt is not None:
+            # Make sure dt is greater than the offset
+            step_start_offset = pybamm.settings.step_start_offset
+            if dt <= step_start_offset:
+                raise pybamm.SolverError(
+                    f"Step time must be at least {pybamm.TimerTime(step_start_offset)}"
+                )
 
-        t_start = old_solution.t[-1]
-        t_end = t_start + dt
-        # Calculate t_eval
-        t_eval = np.linspace(t_start, t_end, npts)
```

Comment:
if `t_eval` is the optional argument then this can be `t_eval = t_eval or np.linspace(t_start, t_end, 2)` or equivalent